program/src/logic.rs

```diff
@@ -382,6 +383,130 @@ pub fn distribute_fees<'a, 'b>(
     Ok(())
 }
 
+/// Checks if the stake accounts on `accounts` correspond to the ones generated
+/// by the validator's seeds. Returns the destination bump seed.
+pub fn check_unstake_accounts(
+    program_id: &Pubkey,
+    lido: &Lido,
+    accounts: &UnstakeAccountsInfo,
+) -> Result<u8, ProgramError> {
+    let validator = lido.validators.get(accounts.validator_vote_account.key)?;
+
+    // If a validator doesn't have a stake account, it cannot be unstaked.
+    if validator.entry.stake_seeds.stake_accounts_seed_begin
+        == validator.entry.stake_seeds.stake_accounts_seed_end
+    {
+        msg!("Attempting to unstake from accounts in a validator that has no stake accounts.");
+        return Err(LidoError::InvalidStakeAccount.into());
+    }
+    // Does not underflow. If `stake_accounts_seed_end == 0` the previous
+    // condition would have failed.
```

Comment:
It could still be that `begin > end` and `end = 0`, but that is disallowed because we require that `end >= begin`. I think this is an invariant that is true at initialization and it is preserved by every instruction that modifies the seeds, and therefore it is true here, but it would be better to just confirm. (Although, if we are in a state where it’s false, then we are in deep trouble anyway.)